crates/consensus/beacon/src/engine/mod.rs

```diff
@@ -793,16 +799,25 @@ where
             EngineSyncEvent::PipelineFinished { result, reached_max_block } => {
                 match result {
                     Ok(ctrl) => {
-                        if ctrl.is_unwind() {
-                            self.sync.set_pipeline_sync_target(current_state.head_block_hash);
-                        } else if reached_max_block {
+                        if reached_max_block {
                             // Terminate the sync early if it's reached the maximum user
                             // configured block.
                             return Some(Ok(()))
                         }
 
+                        let Some(current_state) = self.forkchoice_state else {
+                            warn!(target: "consensus::engine", "No forkchoice state available");
+                            return None
+                        };
+
+                        if ctrl.is_unwind() {
+                            // Attempt to sync to the head block after unwind.
+                            self.sync.set_pipeline_sync_target(current_state.head_block_hash);
+                            return None
+                        }
+
                         // Update the state and hashes of the blockchain tree if possible.
-                        match self.restore_tree_if_possible(*current_state) {
+                        match self.restore_tree_if_possible(current_state) {
```

Comment:
btw I find `restore_` very confusing, because I associate this with error handling/cleanup which is not the case, right?

Reply:
i would say that restore is a good description here because before this call the BlockchainTree is in bad state. hence we attempt to restore it